New testdir-like fixture using pathlib

nicoddemus opened the pull request:
As a step towards getting rid of `py.local`.

This simple experiment shows to me how *painful* this will be to end-users and plugin authors to completely drop this interface: we can see from our own test suite how heavily we rely on `py.path.local`'s interface, for example all `make*` functions return a `py.path` object which test suites often use.

nicoddemus commented on testing/test_pytester.py:
```diff
@@ -799,9 +799,5 @@ def test_parse_summary_line_always_plural():
-    else:
-        p1 = testdir.makepyfile(**{str(absfile): ""})
-        assert str(p1) == (testdir.tmpdir / absfile) + ".py"
+    p1 = testdir.makepyfile(**{str(absfile): ""})
```
This was just testing a behavior discussed in https://github.com/pytest-dev/pytest/pull/6578#discussion_r371035867 and https://github.com/pytest-dev/pytest/pull/6579. I believe the current code is what one expects.

bluetech commented:
It's great to kickstart this effort!

I wonder if we shouldn't attempt to go beyond just the internals and make a Path-based API available externally as well? (If not that's OK, but I think it'd be good to consider anyway).

In general, I'd split the `py.path` dependencies to two:

- Inputs (function arguments, setters): This is the simpler of the two - we can just accept any path and covert to `Path` internally, API is compatible.

- Outputs (return values, properties): This is problematic, not really possible to preserve API in this case.

So if we focus on the outputs problem, I see two approaches:

1. Add differently named functions/properties which return `Path` instead of `py.path.local`. In the case of `Testdir` this would be really unworkable. Which different name will we give the `Path`-equivalent of `testdir.mkdir()`? etc..

2. Add an entirely differently named class/interface which works with `Path`, keep the existing `py.path.local` one as a wrapper around the `Path` one. In the case of `Testdir` it would be probably be `Testpath` and `Testdir` will use a `Testpath` internally.

nicoddemus commented:
> I wonder if we shouldn't attempt to go beyond just the internals and make a Path-based API available externally as well?

Indeed I had the same idea, but thought it better to get the basics rolling first and get feedback, which I did, thanks!

> Add an entirely differently named class/interface which works with Path, keep the existing py.path.local one as a wrapper around the Path one. In the case of Testdir it would be probably be Testpath and Testdir will use a Testpath internally.

Yeah that's how @RonnyPfannschmidt introduced `tmp_path` and `tmp_path_factory`. I will go into this route then.

There's [a project](https://github.com/pytest-dev/pytest/projects/4) to tackle this effort.

nicoddemus commented:
I've introduced `test_path`, which is similar to `testdir` but deals with `Path` objects.

I've changed `Testdir` to be a thin wrapper around `TestPath`, delegating all methods to it and converting results back to `py.path.local` as appropriate. I went that route instead of subclassing because I think this way we have a more explicit API that we are expected to support, plus we don't expose new APIs by accident.

Also here:

* Changed `Testdir.finalize` to private: I don't think any user code uses this explicitly, and if it does it definitely shouldn't.
* Some of the public attributes in `Testdir` became private in `TestPath`, such as `request` and `monkeypatch`, but they are still public in `Testdir` as read-only properties.
* Converted `test_doctest` to use `test_path` and `tmp_path`, to see how the conversion feels like and taste our own medicine: mostly smooth, simple find/replace and a few adjustments here and there (`join` -> `joinpath`, `write` -> `write_text`, etc).

I've kept the commits separate for easy reviewing, but we should squash them when merging.

bluetech commented on doc/en/reference.rst:
```diff
@@ -494,25 +494,27 @@ monkeypatch
 
-testdir
-~~~~~~~
+test_path
```
A `versionadded` would be useful here.

bluetech commented on src/_pytest/pytester.py:
```diff
@@ -379,15 +380,29 @@ def LineMatcher_fixture(request: FixtureRequest) -> "Type[LineMatcher]":
 
 @pytest.fixture
-def testdir(request: FixtureRequest, tmp_path_factory) -> "Testdir":
+def test_path(request: FixtureRequest, tmp_path_factory) -> "TestPath":
```
Dunno why but the name `test_path` feels a bit anemic to me. I do prefer `testpath` -- without the `_` it feels more like a proper noun, rather than a variable name. I'm thinking the name `pytester` itself is also pretty fitting for this. WDYT?

nicoddemus replied:
I agree `test_path` doesn't sit very well, but I'm following our more modern approach of using `_` adequately in names, for example `tmp_path`. As much as I don't like `test_path` very much, I would rather prefer consistency... but not 100% sure either.

bluetech replied:
Right, consistency with `temp_path` means it should be `test_path`, not `testpath` (for the same reason above I would have preferred if it were `tmppath` but that's irrelevant now).

The name `pytester` still works though, unless you don't like it of course.

nicoddemus replied:
> The name pytester still works though, unless you don't like it of course.

Sure, but I'm not sure if you want me to do anything here though, as `pytester` is the name of the module and I didn't touch that.

nicoddemus replied:
>  I'm thinking the name pytester itself is also pretty fitting for this. WDYT?

OH you are suggesting naming the fixture `pytester`, sorry missed that bit! 😆 

That's wild, but I definitely agree: the fixture is much more about running and checking tests than a "directory containing tests". 😁 

@RonnyPfannschmidt what do you think?

RonnyPfannschmidt replied:
Im not opposed to a name change, but I do wonder about turning it into a composition

The-Compiler replied:
FWIW :+1: on naming the fixture `pytester`. I was always confused about the plugin being called `pytester` but the fixture being `testdir`, and I agree it's too common/vague as a name (it could be a directory containing test data or something). Seems like the right way forward, and hopefully we can then deprecate/remove `testdir` at some point in the (far) future?

bluetech commented on src/_pytest/pytester.py:
```diff
@@ -379,15 +380,29 @@ def LineMatcher_fixture(request: FixtureRequest) -> "Type[LineMatcher]":
 
-    It is particularly useful for testing plugins. It is similar to the `tmpdir` fixture
-    but provides methods which aid in testing pytest itself.
+    It provides an empty folder (accessible through the ``tmp_path`` attribute) where pytest
```
```suggestion
    It provides an empty directory (accessible through the ``tmp_path`` attribute) where pytest
```

bluetech commented on src/_pytest/pytester.py:
```diff
@@ -550,12 +565,15 @@ def restore(self) -> None:
+    the current  working directory to ``tmp_path`` and environment variables during initialization.
 
     Attributes:
 
```
This commit should remove the "For backward compatibility, the read-only property ``tmpdir`` returns" comment below (can't comment on it directly)

bluetech commented on src/_pytest/pytester.py:
```diff
@@ -676,12 +692,11 @@ def to_text(s):
                 ret = p
-        if ret is not None:
-            ret = py.path.local(ret)
+        assert ret is not None
```
I wasn't brave enough to add this assert, but it eliminates all of the return types of the `make*` methods so :+1: from me.

bluetech commented on src/_pytest/pytester.py:
```diff
@@ -829,21 +844,21 @@ def copy_example(self, name=None) -> py.path.local:
-            # we might need to roll our own
-            py.path.local(str(example_path)).copy(self.tmpdir)
-            return py.path.local(str(self.tmp_path))
+            # we will need to roll our in order to drop py.path.local completely
```
```suggestion
            # we will need to roll our own in order to drop py.path.local completely
```

bluetech commented:
Also this has some conflicts with the "Some minor fixes & type annotations" PR I just merged -- let me know if it's anything problematic.

nicoddemus commented:
> Seems like the right way forward, and hopefully we can then deprecate/remove testdir at some point in the (far) future?

Yep that's the idea. 👍

nicoddemus commented:
Closed in favor of https://github.com/pytest-dev/pytest/pull/7854.
